src/wasm/wasm-ir-builder.cpp

```diff
@@ -634,18 +758,42 @@ Result<> IRBuilder::visitEnd() {
     iff->ifFalse = *expr;
     iff->finalize(iff->type);
     push(maybeWrapForLabel(iff));
+  } else if (auto* tryy = scope.getTry()) {
+    tryy->body = *expr;
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
+  } else if (Try * tryy;
+             (tryy = scope.getCatch()) || (tryy = scope.getCatchAll())) {
+    tryy->catchBodies.push_back(*expr);
+    tryy->finalize(tryy->type);
+    push(maybeWrapForLabel(tryy));
   } else {
     WASM_UNREACHABLE("unexpected scope kind");
   }
   return Ok{};
 }
 
-Result<Index> IRBuilder::getLabelIndex(Name label) {
+Result<Index> IRBuilder::getLabelIndex(Name label, bool inDelegate) {
   auto it = labelDepths.find(label);
   if (it == labelDepths.end() || it->second.empty()) {
-    return Err{"unexpected label '"s + label.toString()};
+    return Err{"unexpected label '"s + label.toString() + "'"};
   }
-  return scopeStack.size() - it->second.back();
+  auto index = scopeStack.size() - it->second.back();
+  if (inDelegate) {
+    if (index == 0) {
+      // The real label we're referencing, if it exists, has been shadowed by
+      // the `try`. Get the previous label with this name instead.
```

Comment:
Can you add a comment with a very short code snippet or something to show what this case is?

Reply:
Done 👍